flashlight/fl/nn/modules/Container.h

```diff
@@ -117,6 +167,47 @@ class FL_API Container : public Module {
   virtual std::string prettyString() const override;
 };
 
+/**
+ * Adds a copy constructor, copy assignment operator, move constructor, move
+ * assignment operator and clone method to the class. This should only be used
+ * if the class basically acts as a wrapper around a container such that no
+ * custom module ownership is used. Users should implement these methods
+ * themselves if any custom ownership is utilised.
+ * The following is an example of custom ownership, where a shared_ptr is used
+ * to share ownership between the base Container class an MyContainer class
+ * such that they need to be manually syncronised when performing a copy.
+ * \code
+    class MyContainer : public Container {
+    public:
+      MyContainer() {
+        lin_ = std::make_shared<Linear>(10, 20);
+        add(lin_);
+      }
+      std::shared_ptr<Linear> lin_;
+    };
+   \endcode
+ */
+#define FL_BASIC_CONTAINER_CLONING(ContainerClass)             \
+  ContainerClass(const ContainerClass& other) {                \
+    train_ = other.train_;                                     \
+    for (auto& mod : other.modules_) {                         \
+      add(mod->clone());                                       \
+    }                                                          \
+  }                                                            \
+  ContainerClass& operator=(const ContainerClass& other) {     \
+    train_ = other.train_;                                     \
+    clear();                                                   \
+    for (auto& mod : other.modules_) {                         \
+      add(mod->clone());                                       \
+    }                                                          \
+    return *this;                                              \
+  }                                                            \
+  ContainerClass(ContainerClass&& other) = default;            \
+  ContainerClass& operator=(ContainerClass&& other) = default; \
+  std::unique_ptr<Module> clone() const override {             \
+    return std::make_unique<ContainerClass>(*this);            \
+  }
```

Comment:
this macro doesn't need a semicolon when used, fwiw

Reply:
Ah yep, fixed.